### DISCUSSION TITLE
Documentation and Tutorial

### DISCUSSION THREAD

**Discussion by panno8M** [godot-nim/.github#2] [Ideas]
We would like to create extensive documentation (on godot-nim specific features) and hangout tutorials to lower the bar for new developers.

For the tutorial, we would like to refer to the “[Your first 2D game](https://docs.godotengine.org/en/stable/getting_started/first_2d_game/index.html)” provided by godot, and would like to include an explanation of the meaning of the syntax, so that the user can proceed with the implementation in turn.

**Comment by panno8M** [comment]
I am not good at English, so I would like to ask @insomniacUNDERSCORElemon to proofread some strange or incomprehensible expressions.
However, I am still in the process of thinking about its structure, so it will be a little further down the road.

**Reply by insomniacUNDERSCORElemon** [reply]
Can do. For what it's worth, everything I've read so far looks fluent and natural to me.

**Reply by panno8M** [reply]
Please tell your compliments to DeepL. I don't know the word "incomprehensible" :P

**Comment by panno8M** [comment]
Document Index (draft)

1. about environment
   1. project set up
      1. usage of wizard (minister)
   1. brief description about gdextension (includes *.gdextension file format)
1. about library
   1. library chart overview
   1. define/register class
   1. spec of gdref
   1. define class method
   1. define class property
   1. use signal
      how to define/call/connect-to signal
   1. introduce sugars
      1. Node/Node
      1. /Singleton
      1. Node as Class 
   1. introduce pragmas
      1. gdsync
      1. signal 
      1. name
      1. getter, setter

**Reply by insomniacUNDERSCORElemon** [reply]
I'm trying to convert my 3.X demo project over, I've started running into issues that are likely part of the change from 3 to 4 (class/name changes). 

For instance I had a command to set the min+max window sizes, but I can't figure out exactly how it wants me to format vector2i. `gdDisplayServer.windowSetMinSize(vector2i(325, 200))` doesn't work (with capital V it says attempting to call routine, vec2 doesn't work like it did in 3.X, using x and y as some other nim things use doesn't seem to help).

**Reply by panno8M** [reply]
This is due to the unfriendly nature of the library.
vector(args..) and vectori(args...) macros are supported, so you can create n-dimensional vectors with them. (Although Godot recognizes up to 4 dimensions)
I like it a bit because you can also use expressions like vectori(vectori(1, 2), 3).

This is certainly a source of confusion. I'll make sugars for each vector type: godot-nim/godotcore#14

https://github.com/godot-nim/demo/blob/0541f61c133968a41fd01dd045954f2612bcd19f/dodge_the_creeps_2d/src/gdMain.nim#L80-L94

**Reply by insomniacUNDERSCORElemon** [reply]
Still no luck on starting from scratch. 

I somehow got to a point where the project build goes fine and the initial language server message is in Godot's output, but nothing happens when running and there are no errors. 

Docs may help here, though an error would be nice (I know I spent a long time without the config.nims file not knowing that was a problem, I don't think there was an error for that either).

**Reply by panno8M** [reply]
Right, the config.nims file is really confusing. I remember trying to integrate it into the *.nimble file and it didn't work.

**Reply by panno8M** [reply]
Please try running it with this change and let me know what you see in the output.

```diff
# config.nims
+ --define: "EnableDebugEvent:on"
+ --define: "EnableDebugInterface:on"
```

**Reply by panno8M** [reply]
@insomniacUNDERSCORElemon 

I have set up a minimal template project in demo, so please compare and see if there are any shortfalls.

**Reply by insomniacUNDERSCORElemon** [reply]
> I have set up a minimal template project in demo, so please compare and see if there are any shortfalls.

You didn't include a scene/node in your template so it doesn't actually work by default (I don't see the echo there either, though it does work via the creeps template). Going with the obvious names seems to not work. That could be part of my issue.

With the debug, no errors pop out but it does see a procedure that I added:

> initialize_scene::process>> process #0...
> Main::contract::virtual::process>> ready...
> Main::contract::procedure::process>> say_what...
> Main::contract::procedure::process>> on_inbox_text_changed...


But it's not getting called for one reason or another. It's probably naming related as stated above but I am not sure where/why.

It's probably some easy mistake to fix with docs, but if you want to take a look at it:
[decbinhex4.zip](https://github.com/user-attachments/files/16398730/decbinhex4.zip)

(note I had main.tscn in the scenes folder but it didn't work there either, echo was more direct before too, not exactly sure if that's the right way to connect signals, and some of those files in src are unused but I'd hope that wouldn't do anything)

**Reply by panno8M** [reply]
Okay, I got it.

1. Main should inherit Control.
2. Unlike gdnative, gdextension recognizes registered types as new classes, not scripts.
   Type “Main” in the `add child node` search window, and you should see the Main class as a subclass of Control (currently Node).

**Reply by panno8M** [reply]
If you simply replace the class to Main while using the current code, it will probably crash. 
This is because the following statement is commented out.

```
if self.getTree.editedSceneRoot ! = nil: return
```

Without this, ready and process would be called even when loading a scene on the editor, for example.

**Reply by panno8M** [reply]
```diff
import godot, strutils
#import godotgen/classes/[gdSceneTree, gdDisplayServer, gdNode, gdInput, gdLineEdit, gdBaseButton, gdLabel]
- import godotgen/classes / [gdInput, gdLineEdit]
+ import godotgen/classes / [gdInput, gdLineEdit, gdSceneTree]
#import godotapi / [gd_os, node_2d, engine, input, control, scene_tree, line_edit, base_button, label]
#import convert

- type Main* = ref object of Node
+ type Main* = ref object of Control
  nonbinary_numbers:set[char] = Digits - {'0', '1'}

proc sayWhat*(self: Main) {.gdsync, name: "say_what".} =
  godot.echo "hello, world!"

method ready(self: Main) {.gdsync.} =
-   # if self.getTree.editedSceneRoot != nil: return
+   if self.getTree.editedSceneRoot != nil: return
  self.sayWhat()

  discard self.getNode("grid/inbox").connect("changed", self.callable "on_inbox_text_changed")

proc on_inbox_text_changed(self: Main, intext:string) {.gdSync.} =
  self.sayWhat()
```

**Comment by panno8M** [comment]
@insomniacUNDERSCORElemon 
I have updated the gdextwiz command and added the new-workspace subcommand.
Can I ask you to update the demo and wiki as well also in order to check the command works?

I will start on the automatic generation of the gdextension file.

**Reply by insomniacUNDERSCORElemon** [reply]
I might be busy today, and I had planned on editing the template into a new one by hand as it seems pretty easy. Also wasn't quite sure on the best approach to updating my own demo, the changes are done but it seems like I might have to redo the changes via the web interface unless I should actually learn the git command (and even then I don't know if it will properly preserve history of the files that change directory).

Either way I'll see what I can do in the next few days. Also I assume the current structure still works at least.

**Reply by panno8M** [reply]
Certainly git is difficult. If you get interested, try touching git. There are plenty of GUI tools, including editor extensions.

In fact, that's enough for me, since some aspects of my motivation are kept alive by discussions with you.

**Reply by insomniacUNDERSCORElemon** [reply]
> in order to check the command works

Command works, though I had to do `gdextwiz new_workspace -n=core`. 

Now that there isn't a nimble file, my old commands don't work.

I had to change the build command in my editor to `cd ../ && nim c bootstrap`. I was hoping I wouldn't still need to change path here (as then different commands are needed depending on given structure), but it seems I do. I know that is tricky with different structures, I would be happy if there were some way to say build current extension and build all (/changed) extensions or something like that.

This is a nitpick, but `switch("out", "$projectdir/lib"/("workspaceentry".toDll))` it'd be nice if that name wasn't OS-specific (I know the command itself isn't, as it spits out a `.so`). I'd say OS_ext, but I know that's weird with camelCase (even gdOs is a bit weird currently, I'd prefer gdOS)

Also `libworkspaceentry` should probably be tied to the workspace name (core->`core.so`), possibly with an optional name defined in gdextwiz (`gdextwiz new_workspace -n=core -lib=myproject`) EDIT: I'll cover that in another thread.

**Reply by panno8M** [reply]
> Command works, though I had to do gdextwiz new_workspace -n=core.

This command should give the following dialog asking for the workspace name even if the options are unspecified, but it doesn't work?

```console
[I] ~/.l/s/g/p/g/dodge_the_creeps_2d (main↑3|✚10) $ gdextwiz new-workspace
wizard new-workspace*     Info: new-workspace is activated 
wizard new-workspace*     Info: project.godot found 
dialog missing-arg#  Warning: workspace name is not specified. 
dialog missing-arg#   Prompt: what's name? 
                enter text ➤  
```

> I had to change the build command in my editor to cd ../ && nim c bootstrap. 

you can use `nim c ../bootstrap` instead.
Here's another reason why I wanted to change from `nimble build` to `nim c`: it feels good to be able to compile a remote project without cd.

> I would be happy if there were some way to say build current extension and build all (/changed) extensions or something like that.

Would the following two commands satisfy this request?

* A command to compile bootstrap found by traversing upward from the current directory.
* A command that traces upward from the current directory and compiles all bootstrap one level down from the project.godot it finds.

**Reply by insomniacUNDERSCORElemon** [reply]
> This command should give the following dialog asking for the workspace name even if the options are unspecified, but it doesn't work?

I was giving a name but it says `new_workspace does not expect non-option arguments at "core"` (or whatever name I give it), thus the `-n=` bit is needed.

> Would the following two commands satisfy this request?

Exactly, yes.

**Reply by panno8M** [reply]
> I was giving a name but it says 

This is expected behavior. When invoked without options, it will begin interacting with the command.

![Peek 2024-08-04 01-05](https://github.com/user-attachments/assets/888938af-0105-4f50-af54-858a068d002d)

**Comment by insomniacUNDERSCORElemon** [comment]
I've added a new page for non-addon structure and added it to the sidebar.

And a PR in .github repo to fix broken links.
